Improve pod deletion resiliency

joestringer opened the pull request:
Review commit by commit.

The main fix is this one:

```
In a user environment, we encountered a scenario where pods would
"disappear" and Cilium seemingly received no notification of deletion
(eg, CNI DELETE command). If this occurs regularly over time, then
this leads to Cilium managing a series of phantom endpoints which no
longer have corresponding k8s objects.

Some symptoms include:

* Exhaustion of IP address management pool, causing inability to deploy
  new endpoints on the node;
* Metrics such as cilium_endpoint_count increase to the maximum
  number of endpoints on the node (typically limited by IP pool);
* Label resolution controllers reporting errors in "cilium status"
  output, such as: pod.core "my-pod-name" not found; and
* Endpoints that Cilium is aware of have no corresponding veth interfaces.

Fix this by periodically iterating the list of exposed endpoints and
checking that the endpoints are still alive and healthy, by checking
that the link is still present on the node. If an endpoint's link is not
present for two consecutive iterations of the garbage collection (and
the endpoint is not otherwise cleaned up by CNI DELETE or similar
operations), then disconnect it from the endpointmanager and release its
resources.
```

I manually tested this new branch by deploying an image based on this
code, deploying an application, finding its corresponding link via the
`cilium endpoint list -o json` API, and deleting it. After two GC intervals,
the controller kicked in to clean up the endpoint.

I also separately checked that if you create a network namespace, then
create a veth and put one half of the veth inside the network namespace,
then delete the network namespace, it causes the veth link to also be
deleted by the kernel. This means that if the container runtime
prematurely deletes the network namespace without informing us via CNI
DELETE, we have a mechanism to detect that the endpoint has gone
away and clean it up.

The new option "endpoint-gc-interval" allows configuration of the
endpoint garbage collection behaviour. If set to less than around 90s,
then it is possible for it to think that an endpoint has gone away and
Cilium missed a CNI delete while a CNI delete is still enqueued to be
sent to Cilium. In this case, Cilium may mistakenly log a warning
regarding the "stray" endpoint despite the upper CNI layer behaving
correctly.

In some datapath modes, most notably when enabling IPVLAN, there is
no host-side network interface to use for this detection so this PR is
ineffective. You may disable the functionality by configuring
`endpoint-gc-interval="0"`.

Related: https://github.com/kubernetes/kubernetes/issues/86944

```release-note
Fix bug where Cilium endpoints are not cleaned up, eventually leading to IP exhaustion despite a small number of endpoints on the node.
```

joestringer committed:
daemon: Refactor endpoint management to expose Delete

Due to some complications around daemon initialization, when deleting an
endpoint, the caller must pass in additional parameters like the daemon,
the IPAM module, and the endpoint manager. This makes it difficult to
share an endpoint deletion implementation with other subsystems.

While it'd be nice to better untangle these objects, doing so could
accrue some risk which we'd rather avoid while preparing an upcoming
bugfix commit. Instead, refactor endpoint management functionality into
a new dedicated type in the daemon package, which hopefully can be
subsequently refactored into more appropriate places (like the
endpointmanager package).

No functional changes, this is purely prepatory work for additional
callers of endpointManager.Delete() from other packages.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on daemon/endpoint.go:
```diff
@@ -1024,3 +1000,51 @@ func (d *Daemon) RemoveRestoredDNSRules(epID uint16) {
+	*endpointmanager.EndpointManager
+
+	d *Daemon
+}
```
Maybe would be better to wrap this up with a wrapper of multiple interfaces, like `pkg/endpoint/endpoing.go#ipReleaser`, and `pkg/endpoint/endpoing.go#monitorOwner` ?

joestringer replied:
I don't think it makes a difference, since `daemon/` has to know about all the concrete types anyway. Abstracting away to use these interfaces is useful in `pkg/` as we can deal with abstract types and not deal directly with the concrete types; or for testing where mocking is important to allow better testing of some core functionality. Neither of those reasons applies here.

Anyhow, this is just a minimal patch to expose the functionality we need. Looking a little more, it seems to me that `endpoint.Delete` should really be `endpointmanager.Delete(ep)` and then `endpointmanager` should have the `ipReleaser` / `monitorOwner` interfaces and actually manage the endpoints instead of passing an endpointmanager into the endpoint package to allow an endpoint to delete itself. If we did this, then I think we can get rid of this new type as it's just making up for the lack of proper structure in the code below it. Maybe there's a safe way to refactor this now but it seems like a better target for the master branch code development rather than a bugfix.

joestringer replied:
^^^ In the latest iteration, this got slightly uglier, but I still didn't want to unravel too many things here. Once we have verified the fix, I plan to follow up on refactoring to clean this up as part of master development of the fix.

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -365,6 +365,14 @@ func (k *K8sWatcher) deletePodHostIP(pod *types.Pod) (bool, error) {
+		// We don't care how many errors occurred, the callee will log them.
+		if errCount := k.endpointManager.DeleteEndpoint(ep); errCount > 0 {
+			return true, fmt.Errorf("encountered %d errors while deleting local pod", errCount)
+		}
```
Some notes here:
1) the entire watcher will be blocked until this function is executed. I would suggest to do this in a go routine, is it possible?
2) In case of an error we are performing an early return and we won't delete the IPs from the ipache (basically the code from line 376 onward.

joestringer replied:
For (2), if the endpoint is local then its corresponding ipcache entry is managed via the `sync-%d-identity-mapping` controller, see `runIPIdentitySync()`. From that I figured that we should either determine that the pod corresponds to a local endpoint and follow the endpoint deletion path, *or*, determine that the pod corresponds to a remote endpoint in which case we must do the ipcache update.

Though that said there's probably no harm in just spinning this out into a separate goroutine and allowing the ipcache delete attempt to occur.

joestringer replied:
I fixed this up to run the logic in a separate goroutine, and added a new "cilium_endpoint_gc_count" metric so we can figure out whether any of these goroutines may be blocked for a long period.

christarazi commented on pkg/metrics/metrics.go:
```diff
@@ -199,6 +199,10 @@ var (
 
+	// EndpointGCCount is a count of goroutines blocking on attempts to
+	// garbage collect an endpoint from the k8s watcher context.
+	EndpointGCCount = NoOpGauge
```
Nit: I think we can be a bit more descriptive on the name to signal that this is an "active" count. Something as simple as `EndpointActiveGCCount` or similar can work. At first glace, `EndpointGCCount` sounds like it's counting how many times we've GC'd endpoints.

joestringer replied:
In the latest iteration, I have renamed this gauge to `cilium_endpoint_active_gc_count`.

I also introduce a new metric `cilium_endpoint_gc_count` which does count the number of times we GC'd endpoints. That way, we can find out if this code is actually effective without having to flood the agent logs.

joestringer replied:
These metrics are now removed.

christarazi reviewed (approved):
LGTM. Minor comment on the metric name. Makes sense to do the minimal refactor here and then continue improving it on master.

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -370,6 +370,79 @@ func (k *K8sWatcher) updatePodHostIP(pod *types.Pod) (bool, error) {
+	// provisioned by Cilium.
+	//
+	// Account for the following cases:
+	// * Endpoint is older than the pod. If we reach this point, we know
```
This isn't 100% reliable. The worker and the controller nodes can be 2 different machines. The creation timestamp for both entities, pod and Cilium endpoint, might be off because the clocks of both machines are not in sync.

joestringer replied:
Hmm, so pod start time is generated at the controller not at kubelet on the node? I figured it would be populated by the node :/

christarazi commented on pkg/k8s/factory_functions.go:
```diff
@@ -593,6 +593,7 @@ func ConvertToPod(obj interface{}) interface{} {
 			SpecHostNetwork:        concreteObj.Spec.HostNetwork,
 			SpecContainers:         containers,
 			StatusPhase:            concreteObj.Status.Phase,
+			StatusStartTime:        concreteObj.Status.StartTime,
```
You also need to add this further below in the `cache.DeletedFinalStateUnknown` case as well

joestringer replied:
I'm following alternative paths of investigation for the core issue at the moment but if I come back to this approach I'll address this.

joestringer committed:
daemon: Plumb the endpoint garbage collector

In a user environment, we encountered a scenario where pods would
"disappear" and Cilium seemingly received no notification of deletion
(eg, CNI DELETE command). If this occurs regularly over time, then
this leads to Cilium managing a series of phantom endpoints which no
longer have corresponding k8s objects.

Some symptoms include:

* Exhaustion of IP address management pool, causing inability to deploy
  new endpoints on the node;
* Metrics such as cilium_endpoint_count increase to the maximum
  number of endpoints on the node (typically limited by IP pool);
* Label resolution controllers reporting errors in "cilium status"
  output, such as: pod.core "my-pod-name" not found; and
* Endpoints that Cilium is aware of have no corresponding veth
  interfaces.

Fix this by periodically iterating the list of exposed endpoints and
checking that the endpoints are still alive and healthy, by checking
that the link is still present on the node. If an endpoint's link is not
present for two consecutive iterations of the garbage collection (and
the endpoint is not otherwise cleaned up by CNI DELETE or similar
operations), then disconnect it from the endpointmanager and release its
resources.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
endpointmanager: Add garbage collection routine

To improve the robustness of Cilium operations in production
environments, introduce a new periodic garbage collection controller
that attempts to detect when endpoints are no longer alive and healthy,
and disconnects them from the agent.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on daemon/endpoint.go:
```diff
@@ -1051,3 +1056,27 @@ func (e *endpointManager) DeleteEndpoint(ep *endpoint.Endpoint) int {
+	// ensure that the caller doesn't unintentionally tear down the
+	// Endpoint thinking that it no longer exists.
+	_, err := netlink.LinkByName(ep.InterfaceName())
+	if _, ok := err.(netlink.LinkNotFoundError); ok {
```
`errors.Is(err, netlink.LinkNotFoundError)`? (this might have been introduced in recent golang versions so the 1.7 branch might not have it)

joestringer replied:
Yeah I went searching for this in the tree and couldn't find any examples in this branch. There are a few tidyups like this and context handling that we can do better in newer branches.

joestringer replied:
Apparently `netlink.LinkNotFoundError` is a struct type, not a singleton error, so this can't be converted to `errors.Is(...)`:

> ../pkg/endpoint/bpf.go:1503:14: type netlink.LinkNotFoundError is not an expression

So I didn't change this for the newer version.

aanm commented on daemon/endpoint.go:
```diff
@@ -1051,3 +1056,27 @@ func (e *endpointManager) DeleteEndpoint(ep *endpoint.Endpoint) int {
+	// currently: That the link has gone missing. Ignore other error to
+	// ensure that the caller doesn't unintentionally tear down the
+	// Endpoint thinking that it no longer exists.
+	_, err := netlink.LinkByName(ep.InterfaceName())
```
just to be sure, this can be done from the Cilium's container?

joestringer replied:
Yes, the Cilium container manages eBPF programs attached to these devices so it must inherently have access to this.

christarazi commented on pkg/endpointmanager/manager.go:
```diff
@@ -75,6 +86,98 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+		log.WithFields(logrus.Fields{
+			logfields.EndpointID:  ep.StringID(),
+			logfields.ContainerID: ep.GetShortContainerID(),
+			"k8sPodName":          ep.GetK8sNamespaceAndPodName(),
```
We don't have a log field for k8s pod names?

joestringer replied:
Apparently not, I lazily copied this from a similar log elsewhere. I can expose a common name for this in the logfields package.

christarazi commented on daemon/endpoint.go:
```diff
@@ -1051,3 +1056,27 @@ func (e *endpointManager) DeleteEndpoint(ep *endpoint.Endpoint) int {
 	return ep.Delete(e.d, e.d.ipam, e, conf)
 }
+
+// Check verifies that endpoints are alive and healthy.
```
With the context loaded in my head, I understand what "alive and healthy" means here. But without the context, it would be vague. Maybe it makes sense to elaborate further? Could simply make the connection in the comment block L1066-1069 or update this godoc

joestringer replied:
Can do so to make this clearer. I guess one of the more important pieces is to link it to the endpointmanager.Checker interface.

christarazi commented on daemon/daemon_main.go:
```diff
@@ -740,6 +740,10 @@ func init() {
 	flags.Int(option.EndpointQueueSize, defaults.EndpointQueueSize, "size of EventQueue per-endpoint")
 	option.BindEnv(option.EndpointQueueSize)
 
+	flags.Duration(option.EndpointGCInterval, 5*time.Minute, "Periodically monitor local endpoint health via link and netns status on this interval and garbage collect them if they become unhealthy, set to 0 to disable")
```
Nit: There's no netns logic specifically AFAICT, so do we need to mention this?

joestringer replied:
:+1: I can drop the netns bit.

I was debating how much detail to put here. On the one hand, if it's vague then it's hard for users to interpret the meaning. On the other hand if it's too specific, it could be wrong if we ended up updating the logic in future for some reason.

christarazi commented on daemon/endpoint.go:
```diff
@@ -1020,10 +1021,14 @@ type endpointManager struct {
+	self := &endpointManager{d: d}
+	mgr := endpointmanager.NewEndpointManager(epSync)
+	epGC := option.Config.EndpointGCInterval
+	if epGC > 0 {
```
Style nit: we can fold this into this if-statement

joestringer replied:
hehe yeah my oldschool desire for <80 character lines ended up with this, it's not strictly necessary.

christarazi reviewed (commented):
First pass through, LGTM. Mostly small nits

joestringer commented on daemon/endpoint.go:
```diff
@@ -1024,3 +1001,82 @@ func (d *Daemon) RemoveRestoredDNSRules(epID uint16) {
+	// ensure that the caller doesn't unintentionally tear down the
+	// Endpoint thinking that it no longer exists.
+	_, err := netlink.LinkByName(ep.InterfaceName())
+	if _, ok := err.(netlink.LinkNotFoundError); ok {
```
This will break ipvlan as in the ipvlan case, the `ep.InterfaceName()` will represent a device inside the container netns, not on the host.

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -75,6 +86,98 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+	mgr.mutex.RLock()
+	defer mgr.mutex.RUnlock()
+
+	// TODO: Consider exposing visibility via Endpoint.SetState().
```
I think it'd be nice to do this, but I'm a little nervous about modifying the endpoint state machine in a bugfix since that can impact a range of things from regeneration handling to metrics for the endpoints. This can also be followed up separately on master.

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -75,6 +86,98 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+	mgr.markedEndpoints = marked
+	mgr.mutex.Unlock()
+
+	// TODO: Handle 'ctx' via DeleteEndpoint path
```
Newer branches have better context handling so I have not handled this here so we don't have to pull in major ctx changes into v1.7.

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -75,6 +86,98 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+			logfields.ContainerID: ep.GetShortContainerID(),
+			"k8sPodName":          ep.GetK8sNamespaceAndPodName(),
+			"URL":                 "https://github.com/kubernetes/kubernetes/issues/86944",
+		}).Warning("Stray endpoint found. You may be affected by upstream Kubernetes issue #86944.")
```
This obviously points to an upstream k8s issue. This case is not guaranteed to be caused by the upstream issue, but the upstream issue could cause this case. As long as the upstream issue remains open, I think it's reasonable to log this in the Cilium logs. If the linked issue is fixed at some point and this log message still triggers, then I think it's reasonable to re-evaluate this log message.

https://github.com/kubernetes/kubernetes/issues/86944

joestringer reviewed (commented):
A few more misc comments FYI.

Unit tests incoming.

christarazi reviewed (approved):
LGTM, clean implementation :rocket:

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -75,6 +86,98 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+	}
+
+	if errors > 0 {
+		return fmt.Errorf("%d errors occurred during endpoint garbage collection", errors)
```
Ah, when we return this to the controller, the controller may run more frequently than once per `endpoint-gc-interval` if it fails and that could lead to additional false-positives as we garbage collect endpoints that were marked quite recently.
